elastic/kibana
# Fix unable to navigate away from 'Timeline' when additional filter is applied

issue: #109802

### Summary
* Fix the redirect bug by using `useLocation`
* Revert "Fix Kibana page crash on redirect navigation when timeline modal is open
* Revert "Fix breadcrumbs path reopens timeline when timeline modal is open 
* update `setBreadcrumbs` to dispatch an "close timeline" action when clicked.

### Quick explanation
After fixing the redirect loop properly, I could revert the previous commits that tried to fix it. I also reverted a fix for closing the modal when breadcrumbs links are clicked because now we can easily dispatch close timeline action without causing a redirect loop.

### Long explanation

We have recurrent issues related to closing the timeline and navigating to other pages.  The current issue was introduced on the following fix attempt:
https://github.com/elastic/kibana/pull/104288

Original issue:
Whenever the redux store changes, `useUrlStateHooks` updates the URL to sync with the store content. But when we navigate to a different page, it is also called to build the new page query string. If we update the store and navigate to a URL in a small time window, it could create an endless loop of `useUrlStateHooks` calls because the current URL doesn't get passed down to the router props. So when calling `history.replace` to update the query string, it would redirect to the previous page.

I tried to fix it before by moving `timeline.isOpen` dispatch to inside `useUrlStateHooks `, so the loop wouldn't happen. But that also creates more bugs. Like the ones described below: 

## Bug one

![Sep-08-2021 11-37-06_low](https://user-images.githubusercontent.com/1490444/132485609-3aef8d31-5033-4d94-80ba-c64c66f9a3bd.gif)

## Bug two
![Sep-08-2021 11-37-40_low](https://user-images.githubusercontent.com/1490444/132485633-efc4a1d7-744e-435b-95bb-b4b01482aaee.gif)


To fix this problem once and for all, I have stepped back. I reverted the two previous fixes, and I am tackling the core issue. When calling `history.replace` to update a query string in the URL, it will use `window.location` instead of the path inside router props. This way, the redirect loop won't happen anymore.




### How to test it

Open the timeline and try to click on every link to any possible page. It should close the timeline and never crash 🙏 


### Checklist
